# P1.
@@ 인공지능은 2x+1=y를 예측할 수 있을까?
* 가중치와 편향값을 각각 하나를 가지는 간단한 모델 

- models : 각 계층의 클래스를 모아 하나의 모델을 만들고 컴파일, 예측, 학습 수행
- layers : 인공신경망의 각 계층을 만드는 클래스 제공

In [12]:
#숫자 5개 중 2개를 학습해 나머지 3개를 예측하는 케라스 예제
import keras
import numpy

#데이터 넣기
x = numpy.array([0,1,2,3,4])
y = x * 2 + 1

#@@ 이미 문제와 답이 주어짐
#print(type(x))
#print(x)
#print(y)
###

#모델링
model = keras.models.Sequential() #model이라는 모델 인스턴스(객체)를 만든다.

model.add(keras.layers.Dense(1, input_shape=(1,))) #모델에 인공지능 계층 추가. Dense 기본 계층.
#입력 노드 하나와 가중치 하나를 가지는 선형 처리 계층

model.compile('SGD', 'mse')#만든 모델을 어떻게 학습할 것인가 + 컴파일
#SGD : 학습 최적화 알고리즘 - 확률적 경사 하강법
#MSE : 손실함수 - 평균제곱오차



#주어진 데이터로 학습 : fit
model.fit(x[:2], y[:2], epochs=100000, verbose=0)
# 미지수는 가중치 하나와 편향값 하나
# 입력 2개
# 출력 2개
# 반복학습횟수 1000
# verbose : 학습 진행 상황 표시



#성능 평가 : predict()
print('Targets:',y[2:])
#이후 3개의 데이터를 출력

print('Predictions:',model.predict(x[2:]).flatten())
#입력값을 주고 출력을 예측


Targets: [5 7 9]
Predictions: [4.999976  6.9999585 8.999941 ]


위의 코드를 돌리면,
>Targets: [5 7 9]

>Predictions: [4.9475074 6.909524  8.87154  ]

    이렇게 출력이 된다.

>Targets: [5 7 9] : y배열의 뒷부분(예측할 y의 답안 - 머신러닝은 x의 세 값(x[2:])을 기반으로 이 세 y값을 정확히 예측해야 한다.)

>Predictions: [4.9475074 6.909524  8.87154  ] : 머신러닝모델(model)이 세 x값을 기반으로 예측한 세 y값

### epoch 한 5만번 돌리면 결과 나오는 듯
Predictions: [4.999976  6.9999585 8.999941 ]
#
그래도 확실한 정수로는 나오지 않는 듯 하다


# > 정리!
------------------
### [ keras 모듈을 이용한 모델 생성을 위한 코드 순서 ]

#### 1) 데이터 마련
1. x 데이터를 마련한다. 
2. y 데이터를 마련한다. (label - 답지) 

#### 2) 모델 생성
3. 모델 빈 껍데기 생성 <br/>
    model = keras.models.Sequential() <br/>
4. 모델에 계층 붙이기 <br/>
    model.add(keras.layers.계층이름()) <br/>
5. 모델 컴파일 <br/>
    model.compile('학습알고리즘', '손실함수') <br/>

#### 3) 모델 학습
6. 모델을 학습한다 <br/>
    model.fit(학습할 데이터 x, 학습할 데이터 y, 학습 반복 수, ...) <br/>

#### 4) 모델 예측 
7. 학습한 모델을 가지고 새로운 데이터의 label의 값을 예측시킨다. <br/>
    model.predict(예측할 데이터 x).flatten() <br/>
    
@@faltten()은 뭐지.